### PR TITLE
Expand terraform to work across the matrix of distributions

### LINKED ISSUE #1778: 2. Expand terraform to work across the matrix of distributions.
The testing for Chef-Infra-Server before release should be run across all the supported distributions.

- Redhat Enterprise Linux 6
- Redhat Enterprise Linux 7
- ~SUSE Linux Enterprise Server 11~ (no longer supported)
- SUSE Linux Enterprise Server 12
- Ubuntu 16.04
- Ubuntu 18.04

We currently have 2 scenarios working for testing on an adhoc basis:
1. standalone upgrade
2. tiered fresh install ipv6

To Test:
If we launch the same scenario with different versions of artifact to be tested, it creates multiple instances in ec2. 
Would running make destroy for a scenario, destroy then all?

### PR DESCRIPTION
**Pull request by snapp**
### Description

- Enable the execution of a scenario multiple times concurrently when passing different values for `ENABLE_IPV6` or `PLATFORM` environment variables.    
- Add `apply-all`, `destroy-all`, and `clean` targets
- Add automatic artifact URL identification and modifies the scenarios to consume an artifact URL and automatically determine what install command to use.

### Issues Resolved

#1778 

### Check List

- [ ] New functionality includes tests
- [ ] All buildkite tests pass
- [ ] Full omnibus build and tests in buildkite pass
- [x] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>
- [ ] PR title is a worthy inclusion in the CHANGELOG

### CODE REVIEW

File: terraform/README.md [hunk 1]
```diff
@@ -44,8 +44,8 @@ Environment variables are used to control how the scenarios are executed and can
 | `AWS_CONTACT` | The primary contact for the resources, this should be the IAM username  and must be able to receive email by appending @chef.io to it (this  person can explain what/why, might not be the business owner) | csnapp |
 | `AWS_SSH_KEY_ID` | The SSH key pair name within the AWS account that you have the private key for. | csnapp |
 | `SCENARIO` | The name of the sub-directory within `scenarios` containing the test you'd like to run. | omnibus-tiered-fresh-install |
-| `UNSTABLE_VERSION` | The version number of the artifact you are testing. | 13.0.40+20190923060037 |
-| `STABLE_VERSION` | The version number of the artifact you want to test upgrades from. | 12.19.31 |
+| `INSTALL_VERSION` | The version number of the artifact you want to install first. | 12.19.31 |
+| `UPGRADE_VERSION` | The version number of the artifact you want to upgrade to. | 13.0.40+20190923060037 |
```
Comment (PrajaktaPurohit) [inline terraform/README.md]: nice!

File: terraform/Makefile [hunk 2]
```diff
@@ -48,6 +49,18 @@ ifneq ($(SCENARIO),)
...
+ifeq ($(WORKSPACE),)
+    ifeq ($(TF_VAR_enable_ipv6), "true")
+        export WORKSPACE = "$(TF_VAR_scenario)-ipv6-$(TF_VAR_platform)"
+    else
+        export WORKSPACE = "$(TF_VAR_scenario)-ipv4-$(TF_VAR_platform)"
+    endif
+else
+	export TF_VAR_scenario = $(shell sed 's/-ipv.-.*//' <<<$(WORKSPACE))
+	export TF_VAR_enable_ipv6 = $(shell grep -q '\-ipv6-' <<<$(WORKSPACE) && echo true || echo false)
+	export TF_VAR_platform = $(shell sed 's/.*-ipv.-//' <<<$(WORKSPACE))
```
Comment (PrajaktaPurohit) [inline terraform/Makefile]: nice!

**Comment by PrajaktaPurohit** [comment]
Really like the work done here. It is very well organized and clean!

File: terraform/README.md [hunk 3]
```diff
@@ -64,12 +64,26 @@ The naming convention for each scenario is `<PACKAGE_TYPE>-<TOPOLOGY>-<INSTALL_S
...
 ***NOTE:*** **There is no automatic reaping of the scenario.**
 
+## Working with Active Scenarios
+
+### List Active Scenarios
+
+For terraform to track multiple concurrent scenarios it uses a concept called a `workspace`.
+
+The `workspace` name is the combination of the following variables `SCENARIO-ENABLE_IPV6-PLATFORM` (e.g. `omnibus-tiered-fresh-install-ipv6-rhel-7`)
+
```
Comment (PrajaktaPurohit) [inline terraform/README.md]: nice!

**Comment by PrajaktaPurohit** [comment]
some output for context:
Running all the scenarios:
```➜  terraform git:(csnapp/1778) INSTALL_VERSION=12.19.31 UPGRADE_VERSION=13.0.40 make apply-all
^[[6~-e
Applying omnibus-standalone-upgrade-from-stable-ipv6-ubuntu-18.04:
	output log can be found in /tmp/omnibus-standalone-upgrade-from-stable-ipv6-ubuntu-18.04.log)
-e
Applying omnibus-tiered-fresh-install-ipv6-ubuntu-18.04:
	output log can be found in /tmp/omnibus-tiered-fresh-install-ipv6-ubuntu-18.04.log)
-en
Waiting for scenarios to complete ...
Done
```
Listing the active workspaces:
```➜  terraform git:(csnapp/1778) make list-active-scenarios
omnibus-standalone-upgrade-from-stable-ipv4-rhel-7
omnibus-standalone-upgrade-from-stable-ipv6-ubuntu-18.04
omnibus-tiered-fresh-install-ipv6-ubuntu-18.04
omnibus-tiered-fresh-install-ipv4-rhel-7
```
Running different configurations on other distributions
```➜  terraform git:(csnapp/1778)  ENABLE_IPV6=false PLATFORM=rhel-7 INSTALL_VERSION=12.19.31 UPGRADE_VERSION=13.0.40 make apply-all
-e
Applying omnibus-standalone-upgrade-from-stable-ipv4-rhel-7:
	output log can be found in /tmp/omnibus-standalone-upgrade-from-stable-ipv4-rhel-7.log)
-e
Applying omnibus-tiered-fresh-install-ipv4-rhel-7:
	output log can be found in /tmp/omnibus-tiered-fresh-install-ipv4-rhel-7.log)
-en
Waiting for scenarios to complete ...
Done
```
Destroy a single workspace:
```
SCENARIO=omnibus-standalone-upgrade-from-stable PLATFORM=rhel-7 ENABLE_IPV6=false INSTALL_VERSION=12.19.31 UPGRADE_VERSION=13.0.40 make destroy
Destroying omnibus-standalone-upgrade-from-stable-ipv4-rhel-7

data.http.workstation-ipv4: Refreshing state...
data.aws_vpc.chef_vpc: Refreshing state...
data.aws_ami.ubuntu_1604: Refreshing state...
data.aws_ami.rhel_8: Refreshing state...
data.aws_subnet.chef_subnet: Refreshing state...
data.aws_ami.ubuntu_1804: Refreshing state...
:
:
module.chef_server.aws_security_group.default: Destroying... (ID: sg-042afd14dcd57117a)
module.chef_server.aws_security_group.default: Destruction complete after 1s

Destroy complete! Resources: 3 destroyed.
Switched to workspace "default".
Deleted workspace "omnibus-standalone-upgrade-from-stable-ipv4-rhel-7"!
```
or
```WORKSPACE=omnibus-tiered-fresh-install-ipv4-rhel-7 INSTALL_VERSION=12.19.31 UPGRADE_VERSION=13.0.40 make destroy```
Destroy all
```INSTALL_VERSION=12.19.31 UPGRADE_VERSION=13.0.40 make destroy-all
-e
Destroying omnibus-standalone-upgrade-from-stable-ipv6-ubuntu-18.04
-e

Destroying omnibus-standalone-upgrade-from-stable-ipv6-ubuntu-18.04

data.http.workstation-ipv4: Refreshing state...
-e
Destroying omnibus-tiered-fresh-install-ipv6-ubuntu-18.04
-e

Destroying omnibus-tiered-fresh-install-ipv6-ubuntu-18.04
:
:
Destroy complete! Resources: 3 destroyed.
Switched to workspace "default".
Deleted workspace "omnibus-standalone-upgrade-from-stable-ipv6-ubuntu-18.04"!
module.back_end.aws_instance.default: Destruction complete after 31s
module.back_end.aws_security_group.default: Destroying... (ID: sg-0a0791bbba80b1acf)
module.front_end.aws_instance.default: Destruction complete after 31s
module.front_end.aws_security_group.default: Destroying... (ID: sg-01e68a522f0125866)
module.back_end.aws_security_group.default: Destruction complete after 0s
module.front_end.aws_security_group.default: Destruction complete after 1s

Destroy complete! Resources: 5 destroyed.
Switched to workspace "default".
Deleted workspace "omnibus-tiered-fresh-install-ipv6-ubuntu-18.04"!
Done
```

**Comment by christopher-snapp** [comment]
Just a heads up that I made a minor change to set the default instance type to `t2.medium`.

I originally had it as `t3.medium` but that didn't work with rhel-6, then I changed it to `m4.large` but while that worked, the systems had more resources than they required.

`t2.medium` seems to work and isn't quite as expensive.

**Pull request merged**